src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/BinderHelper.cs

```diff
@@ -402,8 +402,9 @@ internal static CallInfo CreateCallInfo(IEnumerable<CSharpArgumentInfo> argInfos
 
             int argCount = 0;
             List<string> argNames = new List<string>();
-
-            foreach (CSharpArgumentInfo info in argInfos)
+            CSharpArgumentInfo[] infoArray = ToArray(argInfos);
```

Comment:
`ToArray` would make a copy of argInfos if they are already an array. 
Can we have an array here? Would it be ok to not make a copy?   

Assuming that the original code was ok with enumerating twice, it would would not be any worse, I think.

Reply:
I think the possibility of a caller mutating the array after the call is worse (and more likely to actually happen) than the current situation, especially since this moves rather than adds the `ToArray()` so code that did that currently would work fine.

Reply:
Realistically the caller is generated code for the call site and never changes the array, but I see your point. The allocation due to copying is relatively small and is done once per callsite (not per call) and next to everything else is unlikely to matter.

Reply:
Since it was done anyway the allocation is the same as before. The ref write back I do here is ungainly IMO, but it means that the subclass can just cast to the array created here instead of the allocation it was originally doing, so there's still a single array allocation per construction.